screenpipe-audio/src/core.rs

```diff
@@ -159,176 +160,57 @@ async fn get_device_and_config(
-                }
-                // Keep the stream alive until the recording is done
-                while is_running_weak
-                    .upgrade()
-                    .map_or(false, |arc| arc.load(Ordering::Relaxed))
-                {
-                    std::thread::sleep(Duration::from_millis(100));
-                }
-                s.pause().ok();
-                drop(s);
+        while start_time.elapsed() < duration && is_running.load(Ordering::Relaxed) {
+            match tokio::time::timeout(Duration::from_millis(100), receiver.recv()).await {
+                Ok(Ok(chunk)) => collected_audio.extend(chunk),
+                Ok(Err(e)) => error!("error receiving audio data: {}", e),
+                Err(_) => {} // Timeout, continue loop
             }
-            Err(e) => error!("Failed to build input stream: {}", e),
         }
-    });
 
-    info!(
-        "Recording {} for {} seconds",
-        audio_device.to_string(),
-        duration.as_secs()
-    );
-    // Spawn another thread to collect audio data
-    let collector_handle = tokio::spawn(async move {
-        let mut collected_audio = Vec::new();
-        while is_running_weak_4
-            .upgrade()
-            .map_or(false, |arc| arc.load(Ordering::Relaxed))
-        {
-            while let Some(chunk) = audio_queue.pop() {
-                collected_audio.extend(chunk);
+        if !collected_audio.is_empty() {
+            debug!("sending audio segment to audio model");
+            if let Err(e) = whisper_sender.send(AudioInput {
+                data: Arc::new(collected_audio.clone()),
+                device: audio_stream.device.clone(),
+                sample_rate: audio_stream.device_config.sample_rate().0,
+                channels: audio_stream.device_config.channels(),
+            }) {
+                error!("failed to send audio to audio model: {}", e);
             }
-            tokio::time::sleep(Duration::from_millis(10)).await;
-        }
-        collected_audio
-    });
+            debug!("sent audio segment to audio model");
 
-    // Wait for the duration
-    tokio::time::sleep(duration).await;
-
-    // Signal the recording to stop
-    is_running.store(false, Ordering::Relaxed);
-
-    // Wait for the audio thread to finish
-    if let Err(e) = audio_handle.join() {
-        error!("error joining audio thread: {:?}", e);
-    }
-
-    // Collect the final audio data
-    let audio_data = collector_handle.await.unwrap_or_else(|e| {
-        error!("error joining collector thread: {:?}", e);
-        Vec::new()
-    });
-
-    debug!("sending audio to audio model");
-    if let Err(e) = whisper_sender.send(AudioInput {
-        data: Arc::new(audio_data),
-        device: audio_device.clone(),
-        sample_rate,
-        channels,
-    }) {
-        error!("failed to send audio to audio model: {}", e);
+            // Reset collected audio to the last two seconds of recorded audio
```

Comment:
why do we need overlap if streaming audio?

unrelated: i was wondering how we can improve the use case where you have a speaker and the mic and system audio basically duplicate the transcription (probably in this case should only use the system audio), not a priority

Reply:
The chunk my end in the middle of a word. realtime VAD would help with this.